### ISSUE TITLE
MetaMask injecting into XML

### ISSUE THREAD

**Issue by comodoro** [MetaMask/metamask-extension#1398]
I am working on a web application backend (completely unrelated to Ethereum) that returns XML in response to certain queries. I have just set up the environment again and came across something weird: displayed in the first tag in the root tag was a <script>, which clearly has no business being there. It looks like this:

```
<script>
/*global Web3*/ cleanContextForImports() require('web3/dist/web3.min.js') const LocalMessageDuplexStream = require('post-message-stream') 
[... too long ...]
</script>
```contentscript.js:3201 Failed to execute 'postMessage' on 'DOMWindow': The target origin provided ('http://search.beletrik.cz:8088') does not match the recipient window's origin ('null').
PostMessageStream._write @ contentscript.js:3201
doWrite @ contentscript.js:5016
writeOrBuffer @ contentscript.js:5005
Writable.write @ contentscript.js:4943
ondata @ contentscript.js:4088
EventEmitter.emit @ contentscript.js:2439
Readable.read @ contentscript.js:3933
flow @ contentscript.js:4281
resume_ @ contentscript.js:4264
afterTickTwo @ contentscript.js:3237
Item.run @ contentscript.js:3409
drainQueue @ contentscript.js:3379
```

In the console there is this error relating to MetaMask:

```
contentscript.js:3201 Failed to execute 'postMessage' on 'DOMWindow': The target origin provided ('http://search.beletrik.cz:8088') does not match the recipient window's origin ('null').
PostMessageStream._write @ contentscript.js:3201
doWrite @ contentscript.js:5016
writeOrBuffer @ contentscript.js:5005
Writable.write @ contentscript.js:4943
ondata @ contentscript.js:4088
EventEmitter.emit @ contentscript.js:2439
Readable.read @ contentscript.js:3933
flow @ contentscript.js:4281
resume_ @ contentscript.js:4264
afterTickTwo @ contentscript.js:3237
Item.run @ contentscript.js:3409
drainQueue @ contentscript.js:3379

```
Apparently MetaMask thinks it HTML and injects its javascript code. Tthe checking was pointed to me as this code:

https://github.com/MetaMask/metamask-plugin/blob/master/app/scripts/contentscript.js#L67-L91

The URL has no extension and windows.document.doctype is null, so the checking fails. Content-Type is text/xml and the document has no header. Example document is attached as txt.
[dotaz.txt](https://github.com/MetaMask/metamask-plugin/files/985632/dotaz.txt)

I asked a question first on https://ethereum.stackexchange.com/questions/15978/metamask-rewriting-web-pages which led to this bug report.

TL;DR 
Not an issue for me, but one could reasonably expect better checking before injecting javascript code.

**Comment by flyswatter** [comment]
Awesome, thanks for the thorough report!

**Comment by Zanibas** [comment]
I think we will fix this issue once we start injecting on allowed-page basis, rather than a blanket injection on every page.

We've attempted to find reliable ways to detect when to inject web3 and when it's inappropriate, but due to the large number of edge cases, it may very well be easier to let users enable web3 on a per-page basis.

**Comment by flyswatter** [comment]
Since we'll log in per-domain (for iframe support), we probably still need to do xml detection.

**Comment by musically-ut** [comment]
Just wanted to say that this also effects `owl` files: https://protege.stanford.edu/ontologies/pizza/pizza.owl

Perhaps we can use a [`document.documentElement.tagName == "HTML"` check](https://developer.mozilla.org/en-US/docs/Web/API/Document/documentElement)?

**Comment by phra** [comment]
![screenshot from 2017-10-02 15-29-13](https://user-images.githubusercontent.com/984628/31079835-43be6fbc-a787-11e7-95fe-09a2f1855f6f.png)
